### DISCUSSION TITLE
Improvements to onboarding experience

### DISCUSSION THREAD

**Discussion by joelmenezes** [toverainc/willow#172] [Ideas]
I see this from the main page:
```
Docker, building, configuring, flashing, etc is a pain. There are several approaches we plan to take to avoid this and ease the barrier to entry for users to get started.
```

I'm wondering if i can put together something basic and update the readme with screenshots to help with config/flashing.

**Comment by kristiankielhofner** [comment]
Our current highest dev priority is a branch (feature/was) that allows for dynamic configuration. It moves virtually all of the configuration options (other than debug, likely) out of the build process. This is the first step towards ready-to-flash images we will have for the 1.0 release, at which point docker, build, etc will only be for developers who want to further customize the Willow source for their use case.

Frankly we were surprised at the level of interest in Willow at this very early stage, my assumption when very softly announcing it on HN was that a couple dozen (tops) devs who weren't scared away by the fairly technical build and configuration process would be early adopters of Willow.

**Reply by joelmenezes** [reply]
The build process was fairly easy and a couple of hiccups I faced were solved by googling. I've had a couple of instances of willow running full time for the past 2 weeks. 

I think my request is moot at this point so feel free to close this thread.

**Comment by vexter0944** [comment]
Just a thought - have you looked at something like ESPHOME - runs in docker and uses yaml but creates the firmware all in the container via web interface.  Then I download the firmware my pc and flash over usb.  Just tossing the idea out there.

https://hub.docker.com/r/esphome/esphome and https://esphome.io/

**Reply by kristiankielhofner** [reply]
We certainly have - @stintel and I personally use esphome and love it for what it does well!

The issue is that [esphome is very, very, very far away](https://github.com/esphome/feature-requests/issues/2239) from supporting all of the functionality of Willow - the ESP BOX hardware itself, wake word, audio processing, audio streaming (in the optimized way WIS expects), etc.

We know how difficult it is to get all of this functionality to work as well as it does with Willow. Frankly, I'm not sure if it even makes sense to support it in esphome given that full Willow functionality is so far outside of what esphome does, how it does it, and what it is intended for. Of course that's a call for them to make but that issue has been in progress for over a month and even getting basic audio to work has been a challenge even with Willow as a reference (I don't blame them - it's hard).

With our feature/was branch we have "esphome like" functionality in progress with dynamic configuration and over the air updates. I'll note that our mechanisms for dynamic configuration and over the air updates are also significantly faster and lighter as we don't repeat the build steps for updates. We can currently reconfigure devices in less than 10 seconds (almost all of that is reboot + connecting to Wifi). Over the air updates take roughly 15 seconds last I checked.

We will also have hosted and self-hosted browser-based options for initial flash and configuration of Willow devices. With all of this the current difficult and convoluted build process will only apply to Willow developers. This will all be in the Willow 1.0 release, coming early July.

We're excited to see the (likely) much higher adoption of Willow this will bring!

**Reply by vexter0944** [reply]
I might be presuming, but I think you may have misunderstood what I was driving at (if you didn't then I apologize in advance.)  What I was driving at was as a build/onboarding idea for willow - something similar to esphome - such that it's contained in docker, has everything needed to build said firmware and a pretty easy interface to work with.  Then let the container build (like you do today) the firmware and then let me download local and flash vs connecting the BOX to the Docker server and passing in etc.  OTA after the fact will be sweet - and applaud for that effort!  I'm just saying to lower the entry level - maybe even a vagrant or docker container that has a baseline to get the BOX up and running simply (just connecting to wifi perhaps..) and then let your OTA handle stuff with a simple interface or yaml like ESP HOME does.  I might not be explaining myself well here - but hopefully this helps line up what I was thinking.  Keep up the great work - excited to see where this goes!

**Reply by kristiankielhofner** [reply]
I think I'm understanding you:

"We will also have hosted and self-hosted browser-based options for initial flash and configuration of Willow devices. With all of this the current difficult and convoluted build process will only apply to Willow developers. This will all be in the Willow 1.0 release, coming early July."

So onboarding is install Willow Application Server (WAS) docker container - our equivalent of the esphome docker container with web interface. Connect ESP BOX to your computer, go to web interface in browser, configure Willow stuff (Home Assistant/openHAB/etc), enter wifi credentials, click flash. Repeat "click flash" for as many devices as you have.

This is a one-time out of the box onboarding step. From there you can manage connected devices in WAS including configuration and over the air updates:

<img width="717" alt="image" src="https://github.com/toverainc/willow/assets/563162/5d890aab-8f2c-47f5-82e4-711701520c9d">

OTA:

<img width="686" alt="image" src="https://github.com/toverainc/willow/assets/563162/67124bc0-8308-4d51-aedb-704945ddb45a">

**Reply by vexter0944** [reply]
That is SWEET!  Thanks for the response and info - excited!

**Reply by stintel** [reply]
Merged the feature/was branch and tagged [0.1.0-rc.1](https://github.com/toverainc/willow/releases/tag/0.1.0-rc.1). This should greatly improve the on-boarding experience. See the [Quick Start Guide](http://heywillow.io/quick-start-guide/) how to use it.

**Reply by vexter0944** [reply]
@stintel - it did!  Well done implementing this feature!  The ONLY suggestion I'd offer is note that the esp32 box needs to be plugged into your LOCAL pc when flashing. It was not obvious and I tinkered around a fair bit passing in serial to docker before realizing this.   But other than that feedback, I really thought the onboarding was much better now!  Nice work!

**Reply by stintel** [reply]
Glad to hear about your positive experience! We will update the Quick Start Guide with your suggestion, thanks!
